ledger/participant-integration-api/src/main/scala/platform/store/dao/DbDispatcher.scala

```diff
@@ -26,6 +26,19 @@ private[platform] trait DbDispatcher {
   def executeSql[T](databaseMetrics: DatabaseMetrics)(sql: Connection => T)(implicit
       loggingContext: LoggingContext
   ): Future[T]
+
+  def executeSqlEither[E, T](databaseMetrics: DatabaseMetrics)(sql: Connection => Either[E, T])(
+      implicit
+      loggingContext: LoggingContext,
+      executionContext: ExecutionContext,
+  ): Future[Either[E, T]] =
+    executeSql(databaseMetrics) { connection =>
+      sql(connection).left.map { error =>
+        throw new DbDispatcher.LeftException(error)
+      }
+    }.recover { case e: DbDispatcher.LeftException[_] =>
+      Left[E, T](e.error.asInstanceOf[E])
```

Comment:
this is really hacky, and the exception throwing should be reserved normally for error cases, I would also be slightly worried about all that rich exception handling logic until it makes here.
Would be nice if we could achieve the same without exceptions (maybe with calling rollback manually, see my other comment)
What is nice that it is not changing the DbDispatcher implementation: I would definitely keep it that way :+1:

Reply:
thanks for the hint - fixed ;)